# Part 2:  Functions and Control Statements

A distinguishing property of *programming* languages is that the programmer can create their own *functions*.  Creating a *function* is like teaching the computer a new trick.  Typically a function will receive some data as *input*, will perform an *algorithm* involving the input data, and will *output* data when the algorithm terminates.  

In this part, we explore Python functions.  We also explore control statements, which allow a program to behave in different ways for different inputs.  We also introduce the *while loop*, a loop whose repetition can be more carefully controlled than a for loop.  As an application of these techniques, we implement the Euclidean algorithm as a Python function in a few ways, to effectively find the GCD of integers.

At the end, you will be prepared to explore the Collatz conjecture.

## Getting started with Python functions

A *function* in Python is a construction which takes input data, performs some actions, and outputs data.  It is best to start with a few examples and break down the code.  Here is a function `square`.  Run the code as usual by pressing *shift-Enter* when the code block is selected.

In [1]:
def square(x):
    answer = x * x
    return answer

To see "under the hood" what Python is doing, we import the Python disassembler.

In [2]:
from dis import dis

In [3]:
dis(square)

  2           0 LOAD_FAST                0 (x)
              2 LOAD_FAST                0 (x)
              4 BINARY_MULTIPLY
              6 STORE_FAST               1 (answer)

  3           8 LOAD_FAST                1 (answer)
             10 RETURN_VALUE


When you run the code block, you probably didn't see anything happen.  But you have effectively taught your computer a new trick, increasing the vocabulary of commands it understands through the Python interpreter.  

More specifically, Python has turned your neat "square function" into a series of very quickly runnable commands ("LOAD_FAST" and "BINARY_MULTIPLY").  That way, every time you want to use your square function, Python will just go through the same series of quick operations.

You don't really need to know the "disassembled" square function above... you can use the `square` command as you wish.

In [4]:
square(12)

144

In [5]:
square(1.5)

2.25

Let's break down the syntax of the *function declaration*, line by line.

```python
def square(x):
    answer = x * x
    return answer
```

The first line begins with the Python reserved word `def`.  (So don't use `def` as a variable name!).  The word `def` stands for "define" and it defines a function called `square`.  After the function name `square` comes parentheses `(x)` containing the **argument** `x`.  The *arguments* or *parameters* of a function refer to the input data.  Even if your function has no arguments, you need parentheses.  The argument `x` is used to name whatever number is input into the `square` function.  

At the end of the function declaration line is a colon `:` and the following two lines are indented.  As in the case of for loops, the colon and indentation are signals of *scope*.  Everything on the indented lines is considered within the *scope of the function* and is carried out when the function is used later.

The second line `answer = x * x` is the beginning of the scope of the function.  It declares a variable `answer` and sets the value to be `x * x`.  So if the argument `x` is 12, then `answer` will be set to 144.  The variable `answer`, being declared within the scope of the function, will not be accessible outside the scope of the function.  It is called a **local variable**.

The last line `return answer` contains the Python reserved word `return`, which terminates the function and outputs the value of the variable `answer`.  So when you apply the function with the command `square(1.5)`, the number `1.5` is `passed` as the argument `x`, and `answer` is `2.25`, and that number `2.25` becomes the output.

A function does not have to return a value.  Some functions might just provide some information.  Here is a function which displays the result of division with remainder as a sentence with addition and multiplication.

In [7]:
def display_divmod(a,b):
    quotient = a // b # Integer division
    remainder = a % b #
    print("{} = {} ({}) + {}".format(a,quotient,b,remainder))

If you want a quick look under the hood, here is the function disassembled.  Again, you don't need to know what everything means.  The four-line code above is translated into 20 very fast steps below, numbered 0,2,4, up to 38.

In [8]:
dis(display_divmod)

  2           0 LOAD_FAST                0 (a)
              2 LOAD_FAST                1 (b)
              4 BINARY_FLOOR_DIVIDE
              6 STORE_FAST               2 (quotient)

  3           8 LOAD_FAST                0 (a)
             10 LOAD_FAST                1 (b)
             12 BINARY_MODULO
             14 STORE_FAST               3 (remainder)

  4          16 LOAD_GLOBAL              0 (print)
             18 LOAD_CONST               1 ('{} = {} ({}) + {}')
             20 LOAD_METHOD              1 (format)
             22 LOAD_FAST                0 (a)
             24 LOAD_FAST                2 (quotient)
             26 LOAD_FAST                1 (b)
             28 LOAD_FAST                3 (remainder)
             30 CALL_METHOD              4
             32 CALL_FUNCTION            1
             34 POP_TOP
             36 LOAD_CONST               0 (None)
             38 RETURN_VALUE


Below, we can run the function to see the output.

In [9]:
display_divmod(23,5)

23 = 4 (5) + 3


Notice that this function has no `return` line.  The function terminates automatically at the end of its scope.

The function also uses Python's **string formatting**.  This has changed between Python 2.x and 3.x, and this notebook uses Python 3.x syntax.

String formatting allows you to insert placeholders like `{}` within a string, and later fill those places with a list of things.  

In [10]:
print("My favorite number is {}".format(17))  # The .format "method" substitutes 17 for {}

My favorite number is 17


In [11]:
print("{} + {} = {}".format(13,12,13+12))

13 + 12 = 25


The `format` command is an example of a **string method**.  It has the effect of replacing all placeholders `{}` by the its inputs, in sequence.  There is an intricate syntax for these placeholders, to allow one to match placeholders with values in different orders, and to format different kinds of values.  Here is the [official reference for string formatting in Python 3.x](https://docs.python.org/3/library/string.html#formatstrings).  We will only use the most basic features, exhibited below.

In [12]:
print ("The number {} comes before {}.".format(1,2)) # This should be familiar.
print ("The number {1} comes before {0}.".format(1,2)) # What happens?
print ("The number {1} comes before {1}.".format(1,2)) # Got it now?



The number 1 comes before 2.
The number 2 comes before 1.
The number 2 comes before 2.


- {0} is the first number, while {1} is the second number in the format.

By placing a number in the placeholder, like `{1}`, one can fill in the placeholders with the values in a different order, or repeat the same value.  The format method takes multiple parameters, and they are numbered:  parameter 0, parameter 1, parameter 2, etc..  So the placeholder `{1}` will be replaced by the second parameter (parameter 1).  It's confusing at first, but Python almost always starts counting at zero.

In [13]:
print("pi is approximately {0}".format(3.14159265))
print("pi is approximately {0:f}".format(3.14159265)) # The "f" in "0:f" formats the float.
print("pi is approximately {0:0.3f}".format(3.14159265)) # Choose 3 digits of precision.


pi is approximately 3.14159265
pi is approximately 3.141593
pi is approximately 3.142


If you give some information about how the placeholder is being used, the format method will format things more nicely for printing.  The placeholder `{0:f}` will be replaced by parameter 0, and it will be formatted in a way that is nice for floats (hence the `f`).  Don't try formatting things outside of their type!

In [14]:
print("{:d} is a pretty big integer.".format(2**100)) # d is the formatting code for integers.
print("{:f} is an integer, formatted like a float.".format(2**100))
print("{:f} is a float, of course.".format(1/7))
print("{:s} is a string.".format('Hi there!')) # s is the formatting code for strings.
print("{:d} will give us an error message.".format(1/7))


1267650600228229401496703205376 is a pretty big integer.
1267650600228229401496703205376.000000 is an integer, formatted like a float.
0.142857 is a float, of course.
Hi there! is a string.


ValueError: ignored

In [15]:
from math import sqrt  # Make sure the square root function is loaded.
print("The square root of {0:d} is about {1:f}.".format(1000, sqrt(1000)))

The square root of 1000 is about 31.622777.


### Exercises

1.  What are the signals of scope in Python?

2.  Write a function called area_circle, which takes one argument radius. The function should return the area of the circle, as a floating point number. Then add one line to the function, using string formatting, so that it additionally prints a helpful sentence of the form "The area of a circle of radius 1.0 is 3.14159." (depending on the radius and the area it computes).

3.  Write a function called factorial, which takes one argument called `n`.  The function should return the factorian of `n` when `n` is a positive integer.  Don't worry about what happens when `n` is zero, or a bad input.  Don't use recursion (if you know it) -- just use a for loop.

4. `format` is an example of a "string method".  Another neat one is `replace`.  Try `"Python".replace("yth","arag")` to see what it does.  

5.  Try the formatting codes `%` and `E` (instead of `f`) for a floating point number.  What do they do?

6. Can you think of a reason you might want to have a function with *no* arguments?

In [ ]:
#  Use this space to work on the Exercises.  
#  Remember that you can add a new cell above/below by clicking to the left of a cell,
#  (the cell will have a blue bar at the left) and then pressing "a" or "b" on the keyboard.

**1)** Everything on the indented lines after the colon is considered within the *scope of the function* in Python. 

**2)** Code below:

In [42]:
def area_circle(r):
  print("The area of a circle of radius {:0.1f} is {:0.5f}.".format(r, 3.14159265359 * r * r))

In [45]:
area_circle(1)

The area of a circle of radius 1.0 is 3.14159.


In [46]:
area_circle(2)

The area of a circle of radius 2.0 is 12.56637.


**3)** Code below:

In [128]:
def factorial(n):
  m = n
  if n > 0:
    for i in range(1, n):
      m = m * i
    return m
  elif n == 0:
    print(1)
  else:
    print('undefined')

In [130]:
factorial(3)

6

**4)** The first parameter is replaced with the second parameter, shown in the code below.

In [133]:
print('Python'.replace("yth","arag"))

Paragon


**5)** With the codes below, it looks like result in a float number. `{:%}` results in two more added zeros in the integers with a `%` at the end, but I think this is somehow wrong, and `{:E}` results in an integer or a float times `10 ** exponent`.  

In [174]:
def area_square(r):
  print("The area of a square of a side {:0.1f} is {:5f}.".format(r, r * r))
  print("The area of a square of a side {:0.1f} is {:%}.".format(r, r * r))
  print("The area of a square of a side {:0.1f} is {:E}.".format(r, r * r))

In [178]:
area_square(5)

The area of a square of a side 5.0 is 25.000000.
The area of a square of a side 5.0 is 2500.000000%.
The area of a square of a side 5.0 is 2.500000E+01.


**6)** I can't think of a reason why we might want to have a function with *no* arguments besides the function having a print in the scope with no argument needed. 

## Control statements

It is important for a computer program to behave differently under different circumstances.  The simplest control statements, `if` and its relative `else`, can be used to tell Python to carry out different actions depending on the value of a boolean variable.  The following function exhibits the syntax.

In [179]:
def is_even(n):
    if n%2 == 0:
        print("{} is even.".format(n))
        return True
    else:
        print("{} is odd.".format(n))
        return False

In [180]:
is_even(17)

17 is odd.


False

In [181]:
is_even(1000)

1000 is even.


True

The broad syntax of the function should be familiar.  We have created a function called `is_even` with one argument called `n`.  The body of the function uses the **control statement** `if n%2 == 0:`.  Recall that `n%2` gives the remainder after dividing `n` by `2`.  Thus `n%2` is 0 or 1, depending on whether `n` is even or odd.  Therefore the **boolean** `n%2 == 0` is `True` if `n` is even, and `False` if `n` is odd.

The next two lines (the first `print` and `return` statements) are within the **scope** of the `if <boolean>:` statement, as indicated by the colon and the indentation.  The `if <boolean>:` statement tells the Python interpreter to perform the statements within the scope if the boolean is `True`, and to ignore the statements within the scope if the boolean is `False`.

Putting it together, we can analyze the code.
```python
    if n%2 == 0:
        print("{} is even.".format(n))
        return True
```
If `n` is even, then the Python interpreter will print a sentence of the form `n is even`.  Then the interpreter will return (output) the value `True` and the function will terminate.  If `n` is odd, the Python interpreter will ignore the two lines of scope.

Often we don't just want Python to *do nothing* when a condition is not satisfied.  In the case above, we would rather Python tell us that the number is odd.  The `else:` control statement tells Python what to do in case the `if <boolean>:` control statement receives a `False` boolean.  We analyze the code
```python
    else:
        print("{} is odd.".format(n))
        return False
```
The `print` and `return` commands are within the scope of the `else:` control statement.  So when the `if` statement receives a false signal (the number `n` is odd), the program prints a sentence of the form `n is odd.` and then returns the value `False` and terminates the function.

The function `is_even` is a verbose, or "talkative" sort of function.  Such a function is sometimes useful in an interactive setting, where the programmer wants to understand everything that's going on.  But if the function had to be called a million times, the screen would fill with printed sentences!  In practice, an efficient and silent function `is_even` might look like the following.

In [182]:
def is_even(n):
    return (n%2 == 0)

In [183]:
is_even(17)

False

A `for` loop and an `if` control statement, used together, allow us to carry out a **brute force** search.  We can search for factors in order to check whether a number is prime.  Or we can look for solutions to an equation until we find one.

One thing to note:  the function below begins with a block of text between a triple-quote (three single-quotes when typing).  That text is called a **docstring** and it is meant to document what the function does.  Writing clear docstrings becomes more important as you write longer programs, collaborate with other programmers, and when you want to return months or years later to use a program again.  There are different style conventions for docstrings; for example, here are [Google's docstring conventions](https://google.github.io/styleguide/pyguide.html?showone=Comments#Comments).  We take a less formal approach.

In [184]:
def is_prime(n):
    '''
    Checks whether the argument n is a prime number.
    Uses a brute force search for factors between 1 and n.
    '''
    for j in range(2,n):  # the list of numbers 2,3,...,n-1.
        if n%j == 0:  # is n divisible by j?
            print("{} is a factor of {}.".format(j,n))
            return False
    return True

An important note:  the `return` keyword **terminates** the function.  So as soon as a factor is found, the function terminates and outputs `False`.  If no factor is found, then the function execution survives past the loop, and the line `return True` is executed to terminate the function.

In [185]:
is_prime(91)

7 is a factor of 91.


False

In [186]:
is_prime(101)

True

Try the `is_prime` function on bigger numbers -- try numbers with 4 digits, 5 digits, 6 digits.  Where does it start to slow down?  Do you get any errors when the numbers are large?  Make sure to save your work first, just in case this crashes your computer!  



In [188]:
# Experiment with is_prime here.
is_prime(1234)

2 is a factor of 1234.


False

In [189]:
is_prime(56789)

109 is a factor of 56789.


False

In [190]:
is_prime(123456)

2 is a factor of 123456.


False

In [191]:
is_prime(7891234)

2 is a factor of 7891234.


False

In [192]:
is_prime(56789123)

True

In [193]:
is_prime(456789123)

3 is a factor of 456789123.


False

In [194]:
is_prime(4567891234)

2 is a factor of 4567891234.


False

In [198]:
is_prime(17196591344133151357492400543636671531373187531310283)

449671 is a factor of 17196591344133151357492400543636671531373187531310283.


False

- The function only ever slowed down when the function resulted in `True`, and it does take long or give errors when the number is large. I tried the example in the paragraph below and it didn't give a result so I ended it. 

There are two limiting factors, which we study in more detail later.  These are **time** and **space** (your computer's memory space).  As the loop of `is_prime` goes on and on, it might take your computer a long time!  If each step of the loop takes only a nanosecond (1 billionth of a second), the loop would take about a second when executing `is_prime(1000000001)`.  If you tried `is_prime` on a much larger number, like `is_prime(2**101 - 1)`, the loop would take longer than the lifetime of the Earth.

The other issue that can arise is a problem with *space*.  In Python 3.x, the `range(2,n)` cleverly *avoids* storing all the numbers between `2` and `n-1` in memory.  It just remembers the endpoints, and how to proceed from one number to the next.  In the older version, Python 2.x, the range command `range(2,n)` would have tried to store the entire list of numbers `[2,3,4,...,n-1]` in the memory of your computer.  Your computer has some (4 or 8 or 16, perhaps) gigabytes of memory (RAM).  A gigabyte is a billion bytes, and a byte is enough memory to store a number between 0 and 255.  (More detail about this later!).  So a gigabyte will not even hold a billion numbers.  So our `is_prime` function would have led to memory problems in Python 2.x, but in Python 3.x we don't have to worry (for now) about space.

### Exercises

1.  Create a function `my_abs(x)` which outputs the absolute value of the argument `x`.  (Note that Python already has a built-in `abs(x)` function).  

2.  Modify the `is_prime` function so that it prints a message `Number too big` and returns `None` if the input argument is bigger than one million.  (Note that `None` is a Python reserved word.  You can use the one-line statement `return None`.)  

3.  Write a Python function `thrarity` which takes an argument `n`, and outputs the string `threeven` if `n` is a multiple of three, or `throdd` is `n` is one more than a multiple of three, or `thrugly` if `n` is one less than a multiple of three.  Example:  `thrarity(31)` should output `throdd` and `thrarity(44)` should output `thrugly`.  Hint:  study the `if`/`elif` syntax at [the official Python tutorial](https://docs.python.org/3/tutorial/controlflow.html#if-statements)

4.  Write a Python function `sum_of_squares(n)` which finds and prints a pair of natural numbers $x$, $y$, such that $x^2 + y^2 = n$.  The function should use a brute force search and return `None` if no such pair of numbers $x,y$ exists.  Explore which natural numbers can be expressed as sums of two squares... hint:  look at prime numbers first!

5.  Write a function `gamma(n)` which takes a positive integer n as input, and outputs the difference between the harmonic sum $\frac{1}{1} + \frac{1}{2} + \frac{1}{3} + \cdots + \frac{1}{n}$ and the natural logarithm $\log(n)$.  Use numpy to compute the logarithm, by using the command `from numpy import log` in a separate cell.  Approximate $\gamma(n)$ as $n \rightarrow \infty$.  How large does $n$ need to be to get five digits of precision on this limit?  Can you prove that the limit $\lim_{n \rightarrow \infty} \gamma(n)$ exists?

In [ ]:
#  Use this space for your solutions to the questions.

**1)** Code below:

In [201]:
def my_abs(x):
  if x < 0: 
    x = x * -1
    return x
  else:
    return x

In [209]:
my_abs(-3)

3

In [208]:
my_abs(3)

3

**2)** Code below:

In [210]:
def is_prime(n):
  if n < 1000000000:
    for j in range(2,n):  # the list of numbers 2,3,...,n-1.
        if n%j == 0:  # is n divisible by j?
            print("{} is a factor of {}.".format(j,n))
            return False
    return True
  else:
    return None

In [219]:
is_prime(3000000000)

**3)** Code below:


In [220]:
def thrarity(n):
  if n % 3 == 0:
    print('threeven')
  elif n % 3 == 1:
    print('throdd')
  else:
    print('thrugly')

In [224]:
thrarity(15)

threeven


In [225]:
thrarity(31)

throdd


In [226]:
thrarity(44)

thrugly


**4)** Code below:

In [320]:
def sum_of_squares(n):
  for x in range(1,n):
    x_squared = x * x
    if x_squared == n:
      return True
    else:
      for y in range(1,n):
        y_squared = y * y
        if x_squared + y_squared == n:
          return True
        else:
          continue

In [323]:
sum_of_squares(20)

True

In [322]:
print(sum_of_squares(14))

None


- Below is scratch work for above

In [319]:
n = 101
for x in range(1,n):
  x_squared = x * x
  if x_squared == n:
    print('{} plus zero equals to {}'.format(x_squared,n))
  else:
    for y in range(1,n):
      y_squared = y * y
      if x_squared + y_squared == n:
        print('{} plus {} equals to {}'.format(x_squared,y_squared,n))
      else:
        continue

1 plus 100 equals to 101
100 plus 1 equals to 101


**5)** Code below:

In [330]:
from numpy import log
def gamma(n):
  x = 0
  natural_log = log(n)
  for i in range(1,n+1):
    x += 1 / i
  diff = x - natural_log
  return diff

In [362]:
gamma(100000000)

0.5772156699001876

Since the fucntion above produced  $Σ_{n=1}^{100000000}$$\frac{1}{n}-ln(100000000)=0.5772156699001876$ with little time, one can assume the function goes close to 0.57 or 0.5 or even 0 as $n→∞$. $ln(∞)=undefined$, so using l'hopital's rule, the function becomes $(Σ_{n=1}^{∞}$$-\frac{1}{n^2})-\frac{1}{∞}$, which becomes $Σ_{n=1}^{∞}$$-\frac{1}{n^2}$, which will result in a result.

## Handling errors by raising exceptions.

In the previous batch of exercises, we tried to modify functions to be a bit more intelligent -- identifying when numbers were "too big" for example.  There's a professional way to handle these situations, by raising *exceptions*.  Here is the [official documentation on errors and exceptions](https://docs.python.org/3/tutorial/errors.html).  We will focus on raising exceptions to catch "bad inputs" to functions.  Let's revisit our `is_even` function.

In [363]:
def is_even(n):
    return (n%2 == 0)

In [364]:
is_even(3.14)  # What will this do? It'll result in False.

False

In [365]:
3.14%2  # Well, this explains it!

1.1400000000000001

Although the output of `is_even(3.14)` might be what you want, a smarter function might let the user know that 3.14 should not be input into `is_even`.  We commonly ask whether *integers* are even or odd; if a non-integer ends up input to `is_even`, it might be a sign of a bug elsewhere.  One possibility is to modify the function by manually printing an error message.

In [366]:
def is_even(n):
    if type(n) == int:
        return (n%2 == 0)
    else:
        print("Bad input!  Please input integers only.")
        return None

In [367]:
is_even(4)

True

In [368]:
is_even(3.14)

Bad input!  Please input integers only.


In [369]:
print(is_even(3.14))

Bad input!  Please input integers only.
None


This behavior is a bit better.  The output of the function is neither True nor False, when a non-integer is input.  Instead, the smarter function outputs `None`, which is exactly what it sounds like.

In [370]:
type(None) # A zen command.

NoneType

Instead of manually using a print command and returning None, we can use Python's built-in `exception` class.  Raising exceptions is the Pythonic way of catching errors, and this will make things smoother in the long term.  Here's a new and safe `is_even` function.

In [371]:
def is_even(n):
    if type(n) == int:
        return (n%2 == 0)
    else:
        raise TypeError('Only integers can be even or odd.')

In [372]:
is_even(3)

False

In [373]:
is_even(3.14)

TypeError: ignored

Instead of manually printing the error message and returning `None`, we have raised a `TypeError`.  This gives information about the kind of error, and a custom error message is displayed at the end.  Type errors are meant for situations where a variable belongs to the wrong type.  `TypeError` is just one kind of "exception" -- the full built-in hierarchy of exceptions can be found in the [official Python documentation](https://docs.python.org/3/library/exceptions.html#exception-hierarchy).

Another kind of exception is the `ValueError`.  It seems similar to `TypeError` at first, but `ValueError` is meant to catch an input that has a "bad" value, even if it is the right type.  For example, here is a square root function that only works with positive input.  It should raise an exception (error message) when a negative number is input.  Both positive and negative numbers can be represented as floats, so the error doesn't represent the *wrong type*.  The error represents a *bad value*.  

In [374]:
def sqrt(x):
    '''
    Estimates the square root of a positive number x.
    '''
    if x < 0:
        raise ValueError('Cannot approximate square root of negative numbers.')
    guess= x/2 # A decent place to start
    while True: # A dangerous loop!  See next section...
        new_guess = 0.5 * (guess + x/guess)
        if abs(new_guess - guess) < .000000001: # close enough!
            return new_guess
        guess = new_guess

In [375]:
sqrt(3) # This should be ok.

1.7320508075688772

In [376]:
sqrt(-3)

ValueError: ignored

By raising the `ValueError`, we have avoided an endless loop... the sort of problem that crashes computers!  If you know that your function is only meant for certain kinds of inputs, it is best to catch errors by raising exceptions.

###Exercises

1.  There's a special exception called `ZeroDivisionError`.  When do you think this occurs?  Try to make it occur!  Can you think of a time when you might want to raise this exception yourself?

2.  Make the `is_prime` function safer by raising a `TypeError` or a `ValueError` when a "bad" input occurs.

In [ ]:
# Use this space to work on exercises.

**1)** Anytime a result is divided by a zero, for example, the code below:

In [377]:
print(1/0)

ZeroDivisionError: ignored

**2)** Code below:

In [404]:
def is_prime(n):
  if n < 1:
    raise ValueError('Cannot determine if negative numbers or 0 are prime.')
  while True:
    for j in range(2,n):
      if n % j == 0:
        print("{} is a factor of {}.".format(j, n))
        return False
    return True

In [405]:
is_prime(25)

5 is a factor of 25.


False

In [406]:
is_prime(13)

True

In [407]:
is_prime(-3)

ValueError: ignored

## While loops and implementation of the Eucidean algorithm

We *almost* have all the tools we need to implement the Euclidean algorithm.  The last tool we will need is the **while loop**.  We have seen the *for loop* already, which is very useful for iterating over a range of numbers.  The Euclidean algorithm involves repetition, but there is no way to know in advance how many steps it will take.  The while loop allows us to repeat a process as long as a boolean value (sometimes called a **flag**) is True.  The following countdown example illustrates the structure of a while loop.

In [408]:
def countdown(n):
    current_value = n
    while current_value > 0:  # The condition (current_value > 0) is checked before every instance of the scope!
        print(current_value)
        current_value = current_value - 1
    print("Blastoff!")

In [409]:
countdown(10)

10
9
8
7
6
5
4
3
2
1
Blastoff!


The while loop syntax begins with `while <boolean>:` and the following indented lines comprise the scope of the loop.  If the boolean is `True`, then the scope of the loop is executed.  If the boolean is `True` again afterwards, then the scope of the loop is executed again.  And again and again and so on.

This can be a **dangerous process**!  For example, what would happen if you made a little typo and the last line of the while loop read `current_value = current_value + 1`?  The numbers would increase and increase... and the boolean `current_value > 0` would **always** be `True`.  Therefore the loop would never end.  Bigger and bigger numbers would scroll down your computer screen.  

You might panic under such a circumstance, and maybe turn your computer off to stop the loop.  Here is some advice for when you get stuck in such a neverending loop, and you're using Google Colab.

1.  Back up your work often.  When you're programming, make sure everything else is saved just in case.
2.  Save your programming work (use "Save a copy in GitHub" and/or "Save a copy in Drive") often, especially before running a cell with a loop for the first time.
3.  If you *do* get stuck in a neverending loop, click on "Runtime... Interrupt execution".  This will often unstick the loop and allow you to pick up where you left off.  

Now, if you're feeling brave, save your work, change the while loop so that it never ends, and try to recover where you left off.  But be aware that this could cause your computer to freeze or behave erratically, crashing your browser, etc.  Don't panic... it probably won't break your computer permanently.

The neverending loop causes two problems here.  One is with the computer processor, which will be essentially spinning its wheels.  This is called [busy waiting](https://en.wikipedia.org/wiki/Busy_waiting), and the computer will essentially be busy waiting forever.  The other problem is that your loop is printing more and more lines of text into the notebook.  This could easily crash your web browser, which is trying to store and display zillions of lines of numbers.  So be ready for problems!  

In [411]:
def countdown(n):
    current_value = n
    while current_value > 0:  # The condition (current_value > 0) is checked before every instance of the scope!
        print(current_value)
        current_value = current_value + 1
    print("Blastoff!")

In [412]:
countdown(10)

Streaming output truncated to the last 5000 lines.
97206
97207
97208
97209
97210
97211
97212
97213
97214
97215
97216
97217
97218
97219
97220
97221
97222
97223
97224
97225
97226
97227
97228
97229
97230
97231
97232
97233
97234
97235
97236
97237
97238
97239
97240
97241
97242
97243
97244
97245
97246
97247
97248
97249
97250
97251
97252
97253
97254
97255
97256
97257
97258
97259
97260
97261
97262
97263
97264
97265
97266
97267
97268
97269
97270
97271
97272
97273
97274
97275
97276
97277
97278
97279
97280
97281
97282
97283
97284
97285
97286
97287
97288
97289
97290
97291
97292
97293
97294
97295
97296
97297
97298
97299
97300
97301
97302
97303
97304
97305
97306
97307
97308
97309
97310
97311
97312
97313
97314
97315
97316
97317
97318
97319
97320
97321
97322
97323
97324
97325
97326
97327
97328
97329
97330
97331
97332
97333
97334
97335
97336
97337
97338
97339
97340
97341
97342
97343
97344
97345
97346
97347
97348
97349
97350
97351
97352
97353
97354
97355
97356
97357
97358
97359
97360
97361
97362
97363
9

KeyboardInterrupt: ignored

- I let it run to 100,000.

### The Euclidean algorithm with a while loop

The **Euclidean Algorithm** is a process of repeated division with remainder.  Beginning with two integers `a` (dividend) and `b` (divisor), one computes quotient `q` and remainder `q` to express `a = qb + r`.  Then `b` becomes the dividend and `r` becomes the divisor, and one repeats.  The repetition continues, and the **last nonzero** remainder is the greatest common divisor of `a` and `b`.

We implement the Euclidean algorithm in a few variations.  The first will be a verbose version, to show the user what happens at every step.  We use a while loop to take care of the repetition.

In [464]:
def Euclidean_algorithm(a,b):
    dividend = a
    divisor = b
    while divisor != 0:   # Recall that != means "is not equal to".
        quotient = dividend // divisor
        remainder = dividend % divisor
        print("{} = {} ({}) + {}".format(dividend, quotient, divisor, remainder))
        dividend = divisor  
        divisor = remainder

In [414]:
Euclidean_algorithm(133, 58)

133 = 2 (58) + 17
58 = 3 (17) + 7
17 = 2 (7) + 3
7 = 2 (3) + 1
3 = 3 (1) + 0


In [415]:
Euclidean_algorithm(1312331323, 58123123)

1312331323 = 22 (58123123) + 33622617
58123123 = 1 (33622617) + 24500506
33622617 = 1 (24500506) + 9122111
24500506 = 2 (9122111) + 6256284
9122111 = 1 (6256284) + 2865827
6256284 = 2 (2865827) + 524630
2865827 = 5 (524630) + 242677
524630 = 2 (242677) + 39276
242677 = 6 (39276) + 7021
39276 = 5 (7021) + 4171
7021 = 1 (4171) + 2850
4171 = 1 (2850) + 1321
2850 = 2 (1321) + 208
1321 = 6 (208) + 73
208 = 2 (73) + 62
73 = 1 (62) + 11
62 = 5 (11) + 7
11 = 1 (7) + 4
7 = 1 (4) + 3
4 = 1 (3) + 1
3 = 3 (1) + 0


This is excellent if we want to know every step of the Euclidean algorithm.  If we just want to know the GCD of two numbers, we can be less verbose.  We carefully return the last nonzero remainder after the while loop is concluded.  This last nonzero remainder becomes the divisor when the remainder becomes zero, and then it would become the dividend in the next (unprinted) line.  That is why we return the (absolute value) of the dividend after the loop is concluded.  You might insert a line at the end of the loop, like `print(dividend, divisor, remainder)` to help you track the variables.

In [416]:
def GCD(a,b):
    dividend = a # The first dividend is a.
    divisor = b # The first divisor is b.
    while divisor != 0:   # Recall that != means "not equal to".
        quotient = dividend // divisor
        remainder = dividend % divisor
        dividend = divisor  
        divisor = remainder
    return abs(dividend)  #  abs() is used, since we like our GCDs to be positive.

Note that the `return dividend` statement occurs *after* the scope of the while loop.  So as soon as the *divisor* variable equals zero, the funtion `GCD` returns the *dividend* variable and terminates.

In [417]:
GCD(111,27)

3

In [418]:
GCD(111,-27)

3

We can refine our code in a few ways.  First, note that the `quotient` variable is never used!  It was nice in the verbose version of the Euclidean algorithm, but plays no role in finding the GCD.  Our refined code reads
```python
def GCD(a,b):
    dividend = a  
    divisor = b  
    while divisor != 0:   # Recall that != means "not equal to".
        remainder = dividend % divisor
        dividend = divisor  
        divisor = remainder
    return abs(dividend) 
```

Now there are two slick Python tricks we can use to shorten the code.  The first is called **multiple assignment**.  It is possible to set the values of two variables in a single line of code, with a syntax like below.

In [419]:
x,y = 2,3  # Sets x to 2 and y to 3.

This is particular useful for self-referential assignments, because as for ordinary assignment, the right side is evaluated first and then bound to the variables on the left side.  For example, after the line above, try the line below.  Use print statements to see what the values of the variables are afterwards!

In [420]:
x,y = y,x #  Guess what this does!

In [421]:
print("x =",x) # One could use "x = {}".format(x) too.
print("y =",y)

x = 3
y = 2


Now we can use multiple assignment to turn three lines of code into one line of code.  For the `remainder` variable is only used temporarily before its value is given to the `divisor` variable.  Using multiple assignment, the three lines
```python
    remainder = dividend % divisor
    dividend = divisor  
    divisor = remainder
```
can be written in one line,
```python
    dividend, divisor = divisor, dividend % divisor # Evaluations on the right occur before any assignments!
```

Our newly shortened GCD function looks like this.
```python
def GCD(a,b):
    dividend = a  
    divisor = b  
    while divisor != 0:   # Recall that != means "not equal to".
        dividend, divisor = divisor, dividend % divisor
    return abs(dividend)
```

The next trick involves the while loop.  The usual syntax has the form `while <boolean>:`.  But if `while` is followed by a numerical type, e.g. `while <int>:`, then the scope of the while loop will execute as long as the number is nonzero!  Therefore, the line
```python
while divisor != 0:
```
can be replaced by the shorter line
```python
while divisor:
```

This is truly a trick.  It probably won't speed anything up, and it does not make your program easier to read for beginners.  So use it if you prefer communicating with experienced Python programmers!  Here is the whole function again.
```python
def GCD(a,b):
    dividend = a  
    divisor = b  
    while divisor:   # Executes the scope if divisor is nonzero.
        dividend, divisor = divisor, dividend % divisor
    return abs(dividend)
```

The next optimization is a bit more dangerous for beginners, but it works here.  In general, it can be dangerous to operate directly on the arguments to a function.  But in this setting, it is safe, and makes no real difference to the Python interpreter.  Instead of creating new variables called `dividend` and `divisor`, one can manipulate `a` and `b` directly within the function.  If you do this, the GCD function can be shortened to the following.

In [422]:
def GCD(a,b):
    while b:   # I.e., while b != 0.
        a, b = b, a % b
    return abs(a)

In [426]:
# Try it out.  Try it on some big numbers and see how quick it runs!
GCD(24925,2423)

1

In [427]:
GCD(24893546,28043964)

2

In [431]:
GCD(249735046,57280464)

2

This code is essentially optimal, if one wishes to execute the Euclidean algorithm to find the GCD of two integers.  It almost [matches the GCD code in a standard Python library](https://stackoverflow.com/a/18944210).  It might be slightly faster than our original code -- but there is a tradeoff here between execution speed and readability of code.  In this and the following lessons, we often optimize enough for everyday purposes, but not so much that readability is lost.

### Exercises and explorations

1.  Modify the `is_prime` function by using a while loop instead of `for j in range(2,n):`.  Hint:  the function should contain the lines `j = 2` and `while j < n:` and `j = j + 1` in various places.  Why might this be an improvement from the for loop?  Can you look for factors within a smaller range?

2.  Modify the `Euclidean_algorithm` function to create a function which returns the *number of steps* that the Euclidean algorithm requires, i.e., the number of divisions-with-remainder.  How does the number of steps compare to the size of the input numbers?  

3.  When $a$ and $b$ are integers, $GCD(a,b) \cdot LCM(a,b) = ab$.  Use this fact to write an LCM-function.  Try to make your function output only integers (not floats) and behave in a good way even if $a,b$ are zero.

4.  How does the `GCD(a,b)` function behave when `a` and/or `b` are zero or negative?  Is this good?

5.  Challenge:  Write a function `approximate_e(n)` which approximates $e$ with a maximum error of $10^{-n}$.  (You can assume $n < 1000$ if necessary.)  Try a while loop and `import mpmath`.  Back up your work often! 

In [ ]:
# Use this space to work on the exercises.

**1)** Code below:

In [462]:
def is_prime(n):
  if n < 1:
    raise ValueError('Cannot determine if negative numbers or 0 are prime.')
  else:
    j = 2
    while j < n:
      if n % j == 0:
        print("{} is a factor of {}.".format(j, n))
        return False
      else:
        j = j + 1
        continue
    return True

In [463]:
is_prime(11)

True

- Based on the code above, I don't see how it can be an improvement

**2)** Modify the `Euclidean_algorithm` function to create a function which returns the *number of steps* that the Euclidean algorithm requires, i.e., the number of divisions-with-remainder.  How does the number of steps compare to the size of the input numbers?  

In [520]:
def Euclidean_algorithm(a,b):
  dividend = a
  divisor = b
  n = 1
  while divisor != 0:
    quotient = dividend // divisor
    print("Step {}: We find what multiples with {} to get close to {}, which is {}.".format(n, divisor, dividend, quotient))
    remainder = dividend % divisor
    print("Step {}: Next, we find the remainder by {} - {} ({}), which is {}. This is {} = {} ({}) + {}.".format(n+1, dividend, quotient, divisor, remainder, dividend, quotient, divisor, remainder))
    dividend = divisor
    divisor = remainder
    print("Step {}: Then, we make the divisor the new dividend and the remainder the new divisor, and repeat until the remainder is 0.".format(n+2))
    n = n + 3
  print("The GCD is {}".format(dividend))
  print("Steps required were {}".format(n-2))

In [521]:
Euclidean_algorithm(93,66)

Step 1: We find what multiples with 66 to get close to 93, which is 1.
Step 2: Next, we find the remainder by 93 - 1 (66), which is 27. This is 93 = 1 (66) + 27.
Step 3: Then, we make the divisor the new dividend and the remainder the new divisor, and repeat until the remainder is 0.
Step 4: We find what multiples with 27 to get close to 66, which is 2.
Step 5: Next, we find the remainder by 66 - 2 (27), which is 12. This is 66 = 2 (27) + 12.
Step 6: Then, we make the divisor the new dividend and the remainder the new divisor, and repeat until the remainder is 0.
Step 7: We find what multiples with 12 to get close to 27, which is 2.
Step 8: Next, we find the remainder by 27 - 2 (12), which is 3. This is 27 = 2 (12) + 3.
Step 9: Then, we make the divisor the new dividend and the remainder the new divisor, and repeat until the remainder is 0.
Step 10: We find what multiples with 3 to get close to 12, which is 4.
Step 11: Next, we find the remainder by 12 - 4 (3), which is 0. This is 12 =

In [522]:
def Euclidean_algorithm(a,b):
  dividend = a
  divisor = b
  n = 1
  while divisor != 0:
    quotient = dividend // divisor
    remainder = dividend % divisor
    dividend = divisor
    divisor = remainder
    n = n + 3
  print("The GCD is {}".format(dividend))
  print("Steps required were {}".format(n-2))

In [523]:
Euclidean_algorithm(93,66)

The GCD is 3
Steps required were 11


- Based on the code I wrote, the amount of steps is tripled the size of the input numbers minus one, unless it was supposed to be like the code underneath:

In [528]:
def Euclidean_algorithm(a,b):
  dividend = a
  divisor = b
  n = 1
  while divisor != 0:
    quotient = dividend // divisor
    print("Step {}: We find what multiples with {} to get close to {}, which is {}.".format(n, divisor, dividend, quotient))
    remainder = dividend % divisor
    print("Next, we find the remainder by {} - {} ({}), which is {}. This is {} = {} ({}) + {}.".format(dividend, quotient, divisor, remainder, dividend, quotient, divisor, remainder))
    dividend = divisor
    divisor = remainder
    print("Then, we make the divisor the new dividend and the remainder the new divisor, and repeat until the remainder is 0.")
    n = n + 1
  print("The GCD is {}".format(dividend))
  print("Steps required were {}".format(n-1))

In [529]:
Euclidean_algorithm(93,66)

Step 1: We find what multiples with 66 to get close to 93, which is 1.
Next, we find the remainder by 93 - 1 (66), which is 27. This is 93 = 1 (66) + 27.
Then, we make the divisor the new dividend and the remainder the new divisor, and repeat until the remainder is 0.
Step 2: We find what multiples with 27 to get close to 66, which is 2.
Next, we find the remainder by 66 - 2 (27), which is 12. This is 66 = 2 (27) + 12.
Then, we make the divisor the new dividend and the remainder the new divisor, and repeat until the remainder is 0.
Step 3: We find what multiples with 12 to get close to 27, which is 2.
Next, we find the remainder by 27 - 2 (12), which is 3. This is 27 = 2 (12) + 3.
Then, we make the divisor the new dividend and the remainder the new divisor, and repeat until the remainder is 0.
Step 4: We find what multiples with 3 to get close to 12, which is 4.
Next, we find the remainder by 12 - 4 (3), which is 0. This is 12 = 4 (3) + 0.
Then, we make the divisor the new dividend and

In [530]:
def Euclidean_algorithm(a,b):
  dividend = a
  divisor = b
  n = 1
  while divisor != 0:
    quotient = dividend // divisor
    remainder = dividend % divisor
    dividend = divisor
    divisor = remainder
    n = n + 1
  print("The GCD is {}".format(dividend))
  print("Steps required were {}".format(n-1))

In [531]:
Euclidean_algorithm(93,66)

The GCD is 3
Steps required were 4


- Here, the steps equal to the size of the input numbers, which I think input numbers means the divisor.

**3)**
- Here, I am guessing I need to make a function that returns $LCM(a,b) = \frac{ab}{GCD(a,b)}$, which is the code I have below:

In [591]:
def GCD(a,b):
    dividend = a
    divisor = b
    while divisor != 0:
        quotient = dividend // divisor
        remainder = dividend % divisor
        dividend = divisor  
        divisor = remainder
    return dividend

def LCM(a,b):
  return int(a * b / GCD(a,b))

In [592]:
LCM(3,8)

24

**4)**  How does the `GCD(a,b)` function behave when `a` and/or `b` are zero or negative?  Is this good?

In [548]:
def GCD(a,b):
    dividend = a
    divisor = b
    while divisor != 0:
        quotient = dividend // divisor
        remainder = dividend % divisor
        dividend = divisor  
        divisor = remainder
    return dividend

In [595]:
GCD(-4, 24)

4

In [597]:
def Euclidean_algorithm(a,b):
    dividend = a
    divisor = b
    while divisor != 0:   # Recall that != means "is not equal to".
        quotient = dividend // divisor
        remainder = dividend % divisor
        print("{} = {} ({}) + {}".format(dividend, quotient, divisor, remainder))
        dividend = divisor  
        divisor = remainder

In [600]:
Euclidean_algorithm(-4,45)

-4 = -1 (45) + 41
45 = 1 (41) + 4
41 = 10 (4) + 1
4 = 4 (1) + 0


In [602]:
print(Euclidean_algorithm(5,0))

None


- `a` and/or `b` being zero or negative is good with the `GCD(a,b)` function. 

**5)**  Challenge:  Write a function `approximate_e(n)` which approximates $e$ with a maximum error of $10^{-n}$.  (You can assume $n < 1000$ if necessary.)  Try a while loop and `import mpmath`.  Back up your work often!

In [673]:
def factorial(n):
  m = n
  if n > 0:
    for i in range(1, n):
      m = m * i
    return m
  elif n == 0:
    res = 1
    return res
  else:
    print('undefined')

def approximate_e(n):
  t = 1
  m = 0
  y = n
  while m < y:
    t = t + (1 / factorial(y))
    y = y - 1
  return t

In [677]:
approximate_e(1000)

2.7182818284590455

- I didn't use `import mpmath` because I didn't know how I would use that, so I did it like this and it works.
